### PR TITLE
Components acl

### PR DESCRIPTION
**Pull request by huwshimi**
This branch adds the ability for components to be disabled when in read only mode.

### CODE REVIEW

**Comment by huwshimi** [comment]
This is a rather large branch, I've tried to keep commits tidy so you should probably try and review it commit by commit.

QA:
- load the GUI
- check that you can click around and do everything normally in the inspector, machine view, deployment flow etc.
- open `acl.js`
- change `isReadOnly` to return true
- check that the inputs and buttons etc. are now disabled

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5261/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5262/
Test PASSed.

File: jujugui/static/gui/src/app/components/generic-button/test-generic-button.js [hunk 1]
```diff
@@ -53,6 +53,19 @@ describe('GenericButton', function() {
     assert.equal(callbackStub.callCount, 0);
   });
 
+  it('does submit when disabled', function() {
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/generic-button/test-generic-button.js]: `does not submit`

File: jujugui/static/gui/src/app/components/inspector-config/inspector-config.js [hunk 2]
```diff
@@ -231,6 +232,7 @@ YUI.add('inspector-config', function() {
       var configElements = [];
 
       Object.keys(charmOptions).forEach((key) => {
+        var disabled = this.props.acl.isReadOnly();
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/inspector-config/inspector-config.js]: This doesn't need to be inside the loop as this bool is global. Save some computation here.

File: jujugui/static/gui/src/app/components/inspector-config/inspector-config.js [hunk 3]
```diff
@@ -302,8 +311,12 @@ YUI.add('inspector-config', function() {
           <div className="inspector-config__fields">
             {this._customizeServiceName()}
             <form ref="file-form">
-              <input type="file" ref="file" className="hidden"
-                onChange={this._importConfig} />
+              <input
+                className="hidden"
+                disabled={disabled}
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/inspector-config/inspector-config.js]: Interesting to see if this works.

File: jujugui/static/gui/src/app/components/string-config/string-config.js [hunk 4]
```diff
@@ -61,7 +68,7 @@ YUI.add('string-config', function() {
           <span>{this.props.option.key}{typeString}</span>
           <div
             className="string-config--value"
-            contentEditable="true"
+            contentEditable={!this.props.disabled}
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/string-config/string-config.js]: Like it but may need a disabled class also to style differently

File: jujugui/static/gui/src/app/components/machine-view/machine-view.js [hunk 5]
```diff
@@ -194,15 +197,16 @@ YUI.add('machine-view', function() {
     /**
       Display a list of unplaced units or onboarding.
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/machine-view/machine-view.js]: This description seems wrong.

File: jujugui/static/gui/src/app/components/machine-view/machine/machine.js [hunk 6]
```diff
@@ -199,7 +201,7 @@ YUI.add('machine-view-machine', function() {
       var units = this.props.units.filterByMachine(machine.id, true);
       var menuItems = [{
         label: 'Destroy',
-        action: this._destroyMachine
+        action: !this.props.acl.isReadOnly() && this._destroyMachine
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/machine-view/machine/machine.js]: Does this block the `_destroyMachine` function call?

File: jujugui/static/gui/src/app/components/machine-view/unplaced-unit/unplaced-unit.js [hunk 7]
```diff
@@ -29,6 +29,17 @@ YUI.add('machine-view-unplaced-unit', function() {
...
+    },
+
+    /**
+      Called to check if the component is allowed to be dragged.
+      See: http://gaearon.github.io/react-dnd/docs-drag-source.html
+
+      @method canDrag
+      @param {Object} props The component props.
+    */
+    canDrag: function(props) {
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/machine-view/unplaced-unit/unplaced-unit.js]: Is this function used?

File: jujugui/static/gui/src/app/components/import-export/test-import-export.js [hunk 8]
```diff
@@ -167,4 +178,48 @@ describe('ImportExport', function() {
...
+        hasEntities={true}
+        hideDragOverNotification={sinon.stub()}
+        importBundleFile={sinon.stub()}
+        renderDragOverNotification={sinon.stub()} />, true);
+    var instance = renderer.getMountedInstance();
+    var output = renderer.getRenderOutput();
+    var expected = (
+      <div className="import-export">
+        <span className="import-export__export link"
+          onClick={instance._handleExport}
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/import-export/test-import-export.js]: Shouldn't this be `false`?

File: jujugui/static/gui/src/app/components/inspector-config/test-inspector-config.js [hunk 9]
```diff
@@ -64,14 +70,14 @@ describe('Configuration', function() {
 
     assert.deepEqual(output.props.children[0].props.children[3], [
       <juju.components.StringConfig
-        key="Config-option1"
-        ref="Config-option1"
+        key='Config-option1'
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/inspector-config/test-inspector-config.js]: Why the switch to single quotes here?

**Comment by hatched** [comment]
This is great @huwshimi thanks for making this huge addition!

:+1:

**Comment by anthonydillon** [comment]
Looks good @huwshimi some small comments in code but :+1: if nothing wrong.

File: jujugui/static/gui/src/app/components/generic-button/test-generic-button.js (hunk 1 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/generic-button/test-generic-button.js]: Oops, fixed.

File: jujugui/static/gui/src/app/components/inspector-config/inspector-config.js (hunk 2 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/inspector-config/inspector-config.js]: Good catch.

File: jujugui/static/gui/src/app/components/string-config/string-config.js (hunk 4 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/string-config/string-config.js]: I've added a `--disabled` class too.

File: jujugui/static/gui/src/app/components/machine-view/machine-view.js (hunk 5 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/machine-view/machine-view.js]: Fixed.

File: jujugui/static/gui/src/app/components/machine-view/unplaced-unit/unplaced-unit.js (hunk 7 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/machine-view/unplaced-unit/unplaced-unit.js]: Yeah, it's one of the methods called by the ReactDnD module. They're a bit mysterious as it doesn't get called by any of our code, but rather used internally.

File: jujugui/static/gui/src/app/components/import-export/test-import-export.js (hunk 8 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/import-export/test-import-export.js]: I figured it's probably OK to export in read only mode, so I've only disabled the import button.

File: jujugui/static/gui/src/app/components/inspector-config/test-inspector-config.js (hunk 9 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/inspector-config/test-inspector-config.js]: Oh oops, a find and replace went horribly wrong.

File: jujugui/static/gui/src/app/components/inspector-config/inspector-config.js (hunk 3 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/inspector-config/inspector-config.js]: I think you can have a disabled attribute for files, either way the button that opens this has been disabled too.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5268/
Test PASSed.

**Comment by huwshimi** [comment]
Thanks @hatched and @anthonydillon, I've made those changes and now just need a QA!

**Comment by anthonydillon** [comment]
Thanks a lot Huw. I'll QA this morning once I get to the office.

File: jujugui/static/gui/src/app/components/import-export/test-import-export.js (hunk 8 above)
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/import-export/test-import-export.js]: Your right, good point.

**Comment by anthonydillon** [comment]
Tested everything thing I could in sandbox in both modes and all buttons and controls worked with set to false and all action points are blocked when true.

**Comment by hatched** [comment]
Thanks @huwshimi and @anthonydillon 

:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci.jujugui.org:8080/job/juju-gui-merge/1732

**Comment by hatched** [comment]
That looks like a legitimate test failure that only appears once it's rebased on develop.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5274/
Test PASSed.

**Comment by huwshimi** [comment]
Thanks for the reviews @hatched and @anthonydillon! Greatly appreciated. :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**